docs/design/features/StringDeduplication.md

```diff
@@ -0,0 +1,196 @@
+
+The dedup hash table acts as weak references to the strings. Depending on the scenario we might choose to null out these weak references or not (if it’s more performant to rebuild the hash table). If we we do the former these weak references would be treated as short weak handles so the following will happen before we scan for finalization -
+
+- During BGC final mark phase we will need to null out the strings that are not marked in the hash table. This can be made concurrent. 
+
+- During a full blocking GC we will need to null out the strings that are not marked in the hash table, and relocate the ones that got promoted if we are doing a compacting GC.
+
+**Alternate design points**
+
+- Should we create multiple threads to do the work? 
+
+Deduping can be done leisurely, so it doesn’t merit having multiple threads. 
+
+- Can we use an existing thread to do the work on? 
+
+The finalizer thread is something that’s idling most of the time. However there are already plenty of types of work scheduled to potentially run on the finalizer thread so adding yet another thing, especially an opt in feature, can get messy.
+
+#### **Which strings to dedup**
+
+Only strings allocated on the managed heap will be considered for deduplication. And only the ones that live in the old generations, meaning gen2 an LOH, will be considered. Deduping in young generations would be unproductive as they will likely die very soon. It would be extremely unproductive if the copy we are deduping to actually lived in young gens.
+
+#### **Calculating hash codes and large string consideration**
+
+Currently calling GetHashCode of a string calculates a 32-bit hash code. This is not stored anywhere, unlike the default hash code that’s stored either in the syncblk or a syncblk entry, depending whether the syncblk is also used by something else like locking. As the deduping thread goes through the heap it will calculate the 32-bit hash code and actually install it.
+
+However, a 32-bit hash code means we always need to check for collision by actually comparing the string content if the hash code is the same. And for large strings having to compare the string content could be very costly. For LOH compaction we already allocate a padding object for each large object (which currently takes up at most 0.4% of LOH space on 64-bit). We could make this padding object 1-ptr size larger and store the address of the string it’s deduped too. Likewise we can also use this to store the fact “this is the copy the hash table keeps track of so no need to dedup”. This way we can avoid having to do the detection multiple times for the same string. Below illustrates a scenario where large strings are deduplicated. 
```

Comment:
> We could make this padding object 1-ptr size larger

Making each string object 8 bytes bigger is a lot. It expect that it would eat into the gains we can expect from the deduping in material way.

BTW: Around 2010, we have done quite a bit work to reduce string object by 4 bytes and it had visible impact on important workloads.

Reply:
not every string, just every string on the LOH which is very different :D

Reply:
This text seems to suggest that a 64-bit hash code would be free from collisions. This would not be the case for two reasons: First, even a cryptographically secure hash would have practical collisions at that size. Second, the hash being used will likely be much weaker. It will always be necessary to validate string contents.

Reply:
I'm not sure what caused you to have that impression. we are not calculating 64-bit hash codes at all. the text says "We could make this padding object 1-ptr size larger and store the address of the string it’s deduped too."